src/alloc/isolated_alloc.rs

```diff
@@ -255,18 +285,21 @@ impl IsolatedAlloc {
     /// SAFETY: Same as `dealloc()` with the added requirement that `layout`
     /// must ask for a size larger than the host pagesize.
     unsafe fn dealloc_huge(&mut self, ptr: *mut u8, layout: Layout) {
-        let layout = IsolatedAlloc::huge_normalized_layout(layout, self.page_size);
+        let size = self.huge_normalized_layout(layout);
         // Find the pointer matching in address with the one we got
         let idx = self
             .huge_ptrs
             .iter()
-            .position(|pg| ptr.addr() == pg.0.addr())
+            .position(|pg| ptr.addr() == pg.0.addr().next_multiple_of(layout.align()))
```

Comment:
This logic does not seem right to me... why should this be correct? I think this could return the wrong page! Consider that `pg` here may have been allocated with a smaller alignment, so `next_multiple_of` might return something way outside the range of `pg`.

Please think very slowly and carefully about this, and don't expect the reviewer to find flaws in your logic! The soundness of Miri depends on getting this right.

The right logic should be something like `pg.0.addr() <= ptr.addr() && ptr.addr() < pg.0.addr().strict_add(pg.1)`.

Reply:
> isn't this the exact logic we used when allocating and returning the pointer? 

Yes, so indeed the allocation we want to find will return `true` here. But you also must ensure that all other allocations return `false`. Those may have been allocated with a *different* layout and then your logic goes wrong.

Reply:
Yeah, I realised that eventually x.x